platform/machine/esx/cluster.go

```diff
@@ -0,0 +1,211 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package esx
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+
+	"github.com/coreos/mantle/platform"
+	"github.com/coreos/mantle/platform/api/esx"
+	"github.com/coreos/mantle/platform/conf"
+)
+
+type cluster struct {
+	*platform.BaseCluster
+	api      *esx.API
+	pubIPs   []string
+	privIPs  []string
+	sshKeyID string
+}
+
+func Hosts(cidr string) ([]string, error) {
+	ip, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return nil, err
+	}
+
+	var ips []string
+	for ip := ip.Mask(ipnet.Mask); ipnet.Contains(ip); inc(ip) {
+		ips = append(ips, ip.String())
+	}
+	// remove network address and broadcast address
+	return ips[1 : len(ips)-1], nil
+}
+
+func inc(ip net.IP) {
+	for j := len(ip) - 1; j >= 0; j-- {
+		ip[j]++
+		if ip[j] > 0 {
+			break
+		}
+	}
+}
+
+// NewCluster creates an instance of a Cluster suitable for spawning
+// instances on VMware ESXi hosts.
+//
+// NewCluster will consume the environment variables $ESX_HOST, $ESX_USER,
+// and $ESX_PASSWORD to spawn instances and authenticate.
+//
+// $ESX_PUBLIC_NETWORK_RANGE & $ESX_PRIVATE_NETWORK_RANGE should be given
+// as CIDR blocks (ex: 127.0.0.0/28) and will be used to assign machines a valid
+// network address within the range.
+func NewCluster(opts *esx.Options, rconf *platform.RuntimeConfig) (platform.Cluster, error) {
+	api, err := esx.New(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	bc, err := platform.NewBaseCluster(opts.BaseName, rconf)
+	if err != nil {
+		return nil, err
+	}
+
+	var keyID string
+	if !rconf.NoSSHKeyInMetadata {
+		keys, err := bc.Keys()
+		if err != nil {
+			return nil, err
+		}
+
+		keyID, err = api.AddKey(bc.Name(), keys[0].String())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	pub := Hosts(opts.ESXPublicNetworkRange)
+	for i, ip := range pub {
+		if ip == opts.ESXPublicGateway || ip == opts.ESXServer {
+			pub = append(pub[:i], pub[i+1:])
+		}
+	}
+
+	priv := Host(opts.ESXPrivateNetworkRange)
+	for i, ip := range riv {
+		if ip == opts.ESXPrivateGateway || ip == opts.ESXPrivateServer {
+			priv = append(priv[:i], priv[i+1:])
+		}
+	}
+
+	ac := &cluster{
+		BaseCluster: bc,
+		api:         api,
+		sshKeyID:    keyID,
+		pubIPs:      pub,
+		privIPs:     priv,
+	}
+
+	if !rconf.NoSSHKeyInMetadata {
+		keys, err := ac.Keys()
+		if err != nil {
+			return nil, err
+		}
+
+		if err := api.AddKey(bc.Name(), keys[0].String()); err != nil {
+			return nil, err
+		}
+	}
+
+	return ac, nil
+}
+
+func (ac *cluster) NewMachine(userdata *conf.UserData) (platform.Machine, error) {
```

Comment:
By convention the `cluster` variable should be `ec` for ESX, and likewise `em` for `machine`s.